### PR TITLE
WIP: fido2 support

### PR DESCRIPTION
**Pull request by baloo**
This is still a work in progress, and I'm sharing early, but at some point we should be able to use ctap2 devices.

### CODE REVIEW

**Comment by baloo** [comment]
I still need to plug the protocol bits to the actual HID code, I'll have a look tomorrow.

**Comment by baloo** [comment]
serde_bytes/serde_cbor errors could be related to https://github.com/pyfisch/cbor/issues/102.

File: src/ctap2/commands.rs [hunk 1]
```diff
@@ -0,0 +1,2100 @@
...
+    type Output;
+    fn as_client_pin(&self) -> Result<ClientPIN, Error>;
+    fn parse_response_payload(&self, input: &[u8]) -> Result<Self::Output, Error>;
+}
+
+struct ClientPinResponse {
+    key_agreement: Option<cose::PublicKey>,
+    pin_token: Option<EncryptedPinToken>,
+    /// Number of PIN attempts remaining before lockout.
+    _retries: Option<u8>,
```
Comment (jcjones) [inline src/ctap2/commands.rs]: Is `_retries` really unused? Looks like we always try to set it.

File: src/ctap2/commands.rs [hunk 2]
```diff
@@ -0,0 +1,2100 @@
...
+        let cipher = Cipher::aes_256_cbc();
+
+        // TODO(baloo): This might trigger a panic if size is not big enough
+        let mut output = vec![0; input.len() * 2];
+        output.resize(input.len() * 2, 0);
+        let mut encrypter = Crypter::new(cipher, Mode::Encrypt, self.0.shared_secret(), Some(iv))
+            .map_err(Error::Openssl)?;
+        encrypter.pad(false);
+        let mut out_size = 0;
+        out_size += encrypter.update(input, output.as_mut_slice()
```
Comment (jcjones) [inline src/ctap2/commands.rs]: todo

File: src/ctap2/server.rs [hunk 3]
```diff
@@ -0,0 +1,628 @@
...
+
+    #[deprecated(
+        note = "RelyingParty::Hash is deprecated, ctap2 needs full object with the full id"
+    )]
+    pub fn new_hash(data: &[u8]) -> Result<Self, ()> {
+        Ok(RelyingParty::Hash(RpIdHash::from(data)?))
+    }
+}
+
+// TODO(baloo): should we rename this PublicKeyCredentialUserEntity ?
```
Comment (jcjones) [inline src/ctap2/server.rs]: ...probably

File: src/ctap2/server.rs [hunk 4]
```diff
@@ -0,0 +1,628 @@
...
+            fn visit_map<M>(self, mut map: M) -> Result<Self::Value, M::Error>
+            where
+                M: MapAccess<'de>,
+            {
+                let mut found_type = false;
+                let mut alg = None;
+
+                while let Some(key) = map.next_key()? {
+                    match key {
+                        "alg" => {
```
Comment (jcjones) [inline src/ctap2/server.rs]: this construct that you do over and over should be a macro

File: src/ctap2/server.rs [hunk 5]
```diff
@@ -0,0 +1,628 @@
...
+            display_name: Some(String::from("John P. Smith")),
+        };
+
+        let payload = user.to_ctap2();
+        println!("payload = {:?}", payload);
+        let payload = payload.unwrap();
+        assert_eq!(
+            payload,
+            vec![
+                0xa4, // map(4)
```
Comment (jcjones) [inline src/ctap2/server.rs]: very good

File: src/ctap2/attestation.rs [hunk 6]
```diff
@@ -0,0 +1,675 @@
...
+    pub fn empty() -> Self {
+        AAGuid([0u8; 16])
+    }
+}
+
+impl fmt::Debug for AAGuid {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "AAGuid({:x}{:x}{:x}{:x}-{:x}{:x}-{:x}{:x}-{:x}{:x}-{:x}{:x}{:x}{:x}{:x}{:x})",
```
[the remark this answers is not in the document]
Comment (jcjones) [reply src/ctap2/attestation.rs]: Issue uped

**Comment by jcjones** [comment]
Merging this into unstable branch ctap2. Further comments will be opened as issues against that.

**Pull request merged**